### PR TITLE
Exposes `end_col_offset` attr from python AST

### LINKED ISSUE #12940: Provides `end_col_offset` for mypy ast nodes (pr proposal)
I can work and send a PR solving this

**Feature**
The mypy ast nodes sould have the `end_col_offset` attribute if this information is available in the original python AST nodes

**Pitch**

AST nodes from python version 3.8 and up providess the `end_col_offset` attribute. This information can be used to adress some issues in  mypy:

https://github.com/python/mypy/issues/12288
https://github.com/python/mypy/issues/4868
https://github.com/python/mypy/issues/12513

and also in LibCST

https://github.com/Instagram/LibCST/issues/451

The reason is because the `end_col_offset`  allows to associate a section in the source code with the type information.
We can latter on implement something similar to pyre query API using this feature. 
image

I working on in a TypeVisitor that  represents a MypyFile node as json string (similar to pyre query result), which maybe in the future can be used by `dmypy`. 
But I need `end_col_offest` to be exposed, currently, I have a local copy of mypy that exposes the `end_col_offset`.  I can work in a PR to integrate this modification in mypy. I don't think that PR will change so much stuff.

### PR DESCRIPTION
**Pull request by devmessias**
### Description

Closes #12940

The AST nodes from python versions greater than 3.7 provide the `end_col_offset` attribute. This attribute can be used to solve some issues in mypy.

https://github.com/python/mypy/issues/12288
https://github.com/python/mypy/issues/4868
https://github.com/python/mypy/issues/12513

The AST nodes will have an optional[int] type for `end_col_offset` attributre. 

For python<3.8 this attribute will be None, following the approach already been used in mypy
https://github.com/python/mypy/blob/9ccd081550010700341a46f0e08b8954e5beef70/mypy/fastparse.py#L375

## Test Plan

Currently, the `end_line` attribute is not being tested in mypy so I believe there is no reason to create a test for `end_col_offset`.

### CODE REVIEW

**Comment by devmessias** [comment]
Hi @JelleZijlstra. Could you review this PR?

File: mypy/nodes.py [hunk 1]
```diff
@@ -23,17 +23,19 @@
 
 class Context:
     """Base type for objects that are valid as error message locations."""
-    __slots__ = ('line', 'column', 'end_line')
+    __slots__ = ('line', 'column', 'end_line', "end_column")
```
Comment (JelleZijlstra) [inline mypy/nodes.py:26]:
```suggestion
    __slots__ = ('line', 'column', 'end_line', 'end_column')
```
Keep it consistent

File: mypy/fastparse2.py [hunk 2]
```diff
@@ -599,6 +599,7 @@ def visit_ClassDef(self, n: ast27.ClassDef) -> ClassDef:
         cdef.line = n.lineno + len(n.decorator_list)
         cdef.column = n.col_offset
         cdef.end_line = n.lineno
+        cdef.end_column = n.col_offset
```
Comment (JelleZijlstra) [inline mypy/fastparse2.py:602]: Shouldn't this be None for consistency with `fastparse.py`?

File: mypy/nodes.py [hunk 3]
```diff
@@ -698,10 +707,12 @@ def max_fixed_argc(self) -> int:
...
                  column: Optional[int] = None,
-                 end_line: Optional[int] = None) -> None:
-        super().set_line(target, column, end_line)
+                 end_line: Optional[int] = None,
+                 end_column: Optional[int] = None) -> None:
+        super().set_line(target, column, end_line, end_column)
         for arg in self.arguments:
-            arg.set_line(self.line, self.column, self.end_line)
+            arg.set_line(
+                self.line, self.column, self.end_line, end_column)
```
Comment (JelleZijlstra) [inline mypy/nodes.py:715]:
```suggestion
            arg.set_line(self.line, self.column, self.end_line, end_column)
```
Formatting

**Review by JelleZijlstra — COMMENTED** [review]
Thanks! Two formatting nits and one question.

File: mypy/fastparse2.py (hunk 2 above)
Comment (devmessias) [reply mypy/fastparse2.py:602]: Yes, you're right.

**Comment by devmessias** [comment]
Hi @JelleZijlstra all your suggestions have been implemented.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**